tests/core/util/test_config.py

```diff
@@ -20,47 +32,117 @@
 # log = logging.getLogger(__name__)
 
 
-def write_config(root_path: Path, config: Dict):
+def write_config(
+    root_path: Path,
+    config: Dict,
+    atomic_write: bool,
+    do_sleep: bool,
+    iterations: int,
+    error_queue: Optional[Queue] = None,
+):
     """
     Wait for a random amount of time and write out the config data. With a large
     config, we expect save_config() to require multiple writes.
     """
-    sleep(random.random())
-    # log.warning(f"[pid:{os.getpid()}:{threading.get_ident()}] write_config")
-    # save_config(root_path=root_path, filename="config.yaml", config_data=modified_config)
-    save_config(root_path=root_path, filename="config.yaml", config_data=config)
+    try:
+        for i in range(iterations):
+            # This is a small sleep to get interweaving reads and writes
+            sleep(0.05)
 
+            if do_sleep:
+                sleep(random.random())
+            # log.warning(f"[pid:{os.getpid()}:{threading.get_ident()}] write_config")
+            # save_config(root_path=root_path, filename="config.yaml", config_data=modified_config)
+            if atomic_write:
+                # Note that this is usually atomic but in certain circumstances in Windows it can copy the file,
+                # leading to a non-atomic operation
+                with get_config_lock(root_path, "config.yaml"):
+                    save_config(root_path=root_path, filename="config.yaml", config_data=config)
+            else:
+                path: Path = config_path_for_filename(root_path, filename="config.yaml")
+                # print("Trying to get lock write")
+                with get_config_lock(root_path, "config.yaml"):
+                    # print("Write start")
+                    with tempfile.TemporaryDirectory(dir=path.parent) as tmp_dir:
+                        tmp_path: Path = Path(tmp_dir) / Path("config.yaml")
+                        with open(tmp_path, "w") as f:
+                            yaml.safe_dump(config, f)
+                        shutil.copy2(str(tmp_path), str(path))
```

Comment:
This seems pretty similar to, but not quite the same as `save_config()`.  Should it be the same?  Should `save_config()` just be called here?